File: java/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/GeometryTriangulator.java
package vsdk.toolkit.environment.geometry.geometricProcessing;

import vsdk.toolkit.environment.geometry.Geometry;
import vsdk.toolkit.environment.geometry.surface.FunctionalExplicitSurface;
import vsdk.toolkit.environment.geometry.surface.QuadMesh;
import vsdk.toolkit.environment.geometry.surface.TriangleMesh;
import vsdk.toolkit.environment.geometry.surface.TriangleMeshGroup;

public class GeometryTriangulator {
    public static TriangleMeshGroup exportToTriangleMeshGroup(Geometry geometry)
    {
        if ( geometry == null ) {
            return null;
        }

        if ( geometry instanceof TriangleMeshGroup ) {
            return (TriangleMeshGroup)geometry;
        }
        if ( geometry instanceof TriangleMesh ) {
            TriangleMeshGroup group = new TriangleMeshGroup();
            group.addMesh((TriangleMesh)geometry);
            return group;
        }
        if ( geometry instanceof QuadMesh ) {
            return ((QuadMesh)geometry).exportToTriangleMeshGroup();
        }
        if ( geometry instanceof FunctionalExplicitSurface ) {
            TriangleMeshGroup group = new TriangleMeshGroup();
            group.addMesh(((FunctionalExplicitSurface)geometry).getInternalTriangleMesh());
            return group;
        }

        return null;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/GeometryTriangulator.ts
import { Geometry } from "../Geometry.js";
import { FunctionalExplicitSurface } from "../surface/FunctionalExplicitSurface.js";
import { QuadMesh } from "../surface/QuadMesh.js";
import { TriangleMesh } from "../surface/TriangleMesh.js";
import { TriangleMeshGroup } from "../surface/TriangleMeshGroup.js";

export class GeometryTriangulator {
    public static exportToTriangleMeshGroup(geometry: Geometry | null): TriangleMeshGroup | null {
        if (geometry === null) return null;
        if (geometry instanceof TriangleMeshGroup) return geometry;
        const group = new TriangleMeshGroup();
        if (geometry instanceof TriangleMesh) group.addMesh(geometry);
        else if (geometry instanceof QuadMesh) return geometry.exportToTriangleMeshGroup();
        else if (geometry instanceof FunctionalExplicitSurface) group.addMesh(geometry.getInternalTriangleMesh());
        else return null;
        return group;
    }
}
